[OMM] Remove validation masking, update spec (#374)
microsoft/hlsl-specs: proposals/0024-opacity-micromaps.md

> - `"ConstRayFlags argument of AllocateRayQuery '%0' must be constant"`

Reviewer: ```suggestion
- `"ConstRayFlags argument of AllocateRayQuery '%0' must be constant"`
- `"ConstRayFlags and RayQueryFlags arguments of AllocateRayQuery2 '%0' must be constant"`
```

> Finally, validation will be added to ensure that when `RayFlag::ForceOMM2State`
> is set on any RayQuery Object, that RayQuery object will be accompanied by the
> `RAYQUERY_FLAG::RAYQUERY_FLAG_ALLOW_OPACITY_MICROMAPS` flag.
> 
> Proposed validation error diagnostic:
> 
> - `"RAYQUERY_FLAG::RAYQUERY_FLAG_ALLOW_OPACITY_MICROMAPS is required on constant RayQuery object %0 when RayFlag::ForceOMM2State is set."`.

Reviewer: ```suggestion
Finally, validation will be added for the `AllocateRayQuery2` DXIL operation
to ensure that when `RAY_FLAG_FORCE_OMM_2_STATE` is set on the ConstRayFlags
argument, the `RAYQUERY_FLAG_ALLOW_OPACITY_MICROMAPS` is also set on the
RayQueryFlags argument.

Proposed validation error diagnostic:

- `"RAYQUERY_FLAG_ALLOW_OPACITY_MICROMAPS must be set for RayQueryFlags when RAY_FLAG_FORCE_OMM_2_STATE is set for ConstRayFlags on AllocateRayQuery2 operation %0."`.
```

### Diagnostics

- Check availability-based diagnostics for each flag, including recursing
  through DeclRefs, DeclExprRefs, and their initializers.
  - Check both DXR entry scenarios and non-library RayQuery scenarios.
- Check that any RayQuery object with the `RayFlag::ForceOMM2State` flag
  in its first template argument also has an accompanying 
  `RAYQUERY_FLAG::RAYQUERY_FLAG_ALLOW_OPACITY_MICROMAPS` flag.
- Check diagnostics for subobject, and lack of diagnostics for non-library
  target, where subobjects are ignored.
- Test the custom HLSL availability attribute, that it correctly locates
  values declared in unintuitive ways, through function calls, namespaces,
  etc.

### Validation

> - Check constant flag validation for shader models 6.9 and for an earlier
>   shader model, for each applicable intrinsic.

Reviewer: This should be removed.
```suggestion
```

> - Check subobject flag validation for a shader model less than 6.9.
>   - Compile with the flag to 6.9 and change the target for manual assembly.

Reviewer: This should be removed.
```suggestion
```

> - Check that any RayQuery object with the `RayFlag::ForceOMM2State` flag
>   in its first template argument also has an accompanying 
>   `RAYQUERY_FLAG::RAYQUERY_FLAG_ALLOW_OPACITY_MICROMAPS` flag.  

Reviewer: ```suggestion
- Check constant value validation on `AllocateRayQuery` and `AllocateRayQuery2`
  DXIL ops.
- Check `AllocateRayQuery2` DXIL op validation for
  `RAYQUERY_FLAG_ALLOW_OPACITY_MICROMAPS` requirement when
  `RAY_FLAG_FORCE_OMM_2_STATE` is used.
```
